sno/init.py

```diff
@@ -43,200 +57,368 @@ def import_gpkg(ctx, geopackage, table, list_tables):
-        return dbcur.fetchone()[0]
+        # note: choose FAST method if possible ( i recall them being different with MITAB especially)
+        return self.ogrlayer.GetFeatureCount()
 
     @property
     @functools.lru_cache(maxsize=1)
     def primary_key(self):
-        return gpkg.pk(self.db, self.table)
+        return self.ogrlayer.GetFIDColumn()
 
     @property
     @functools.lru_cache(maxsize=1)
     def geom_cols(self):
-        return gpkg.geom_cols(self.db, self.table)
+        ld = self.ogrlayer.GetLayerDefn()
+        cols = []
+        for i in range(ld.GetGeomFieldCount()):
+            cols.append(ld.GetGeomFieldDefn(i).GetName())
+        return cols
 
     @property
     @functools.lru_cache(maxsize=1)
+    @ungenerator(dict)
     def field_cid_map(self):
-        dbcur = self.db.cursor()
-        q = dbcur.execute(f"PRAGMA table_info({gpkg.ident(self.table)});")
-        return {r["name"]: r["cid"] for r in q}
+        ld = self.ogrlayer.GetLayerDefn()
+
+        yield self.primary_key, 0
+        start = 1
+
+        gc = self.ogrlayer.GetGeometryColumn()
+        if gc:
+            yield gc, 1
+            start += 1
+
+        for i in range(ld.GetFieldCount()):
+            field = ld.GetFieldDefn(i)
+            name = field.GetName()
+            yield name, i + start
+
+    @property
+    @ungenerator(dict)
+    def field_adapter_map(self):
+        ld = self.ogrlayer.GetLayerDefn()
+
+        yield self.primary_key, adapt_value_noop
+
+        gc = self.ogrlayer.GetGeometryColumn()
+        if gc:
+            yield gc, adapt_value_noop
+
+        for i in range(ld.GetFieldCount()):
+            field = ld.GetFieldDefn(i)
+            name = field.GetName()
+            yield name, get_type_value_adapter(field.GetType())
+
+    @ungenerator(dict)
+    def _ogr_feature_to_dict(self, ogr_feature):
+        yield self.primary_key, ogr_feature.GetFID()
+        for name, adapter in self.field_adapter_map.items():
+            if name == self.primary_key:
+                yield name, ogr_feature.GetFID()
+            elif name in self.geom_cols:
+                yield (
+                    self.ogrlayer.GetGeometryColumn(),
+                    gpkg.ogr_to_gpkg_geom(ogr_feature.GetGeometryRef()),
+                )
+            else:
+                value = ogr_feature.GetField(name)
+                yield name, adapter(value)
+
+    def _iter_ogr_features(self):
+        l = self.ogrlayer
+        l.ResetReading()
+        while True:
+            f = l.GetNextFeature()
+            if f is None:
+                # end of iter
+                l.ResetReading()
+                return
+            # Turn an OGRFeature into a name:value dict
+            yield f
+
+    def iter_features(self):
+        for ogr_feature in self._iter_ogr_features():
+            yield self._ogr_feature_to_dict(ogr_feature)
+
+    def iter_features_sorted(self, pk_callback, limit=None):
```

Comment:
what's this actually used for? (I vaguely remember it being related to deterministic imports when we were generating IDs per feature, but we don't do that anymore)

Reply:
good question. I have no idea. The only way to use it from the CLI is via:

```
SNO_IMPORT_OPTIONS='{"iter_func": 2}'
```
Which is confusing. But we test that behaviour, so I couldn't omit it without removing the test. Do you reckon it can be removed?